base/bucket_gocb.go

```diff
@@ -1386,6 +1586,146 @@ func (bucket CouchbaseBucketGoCB) CouchbaseServerVersion() (major uint64, minor
+			}
+		case ViewQueryParamReduce:
+			viewQuery.Reduce(asBool(optionValue))
+		case ViewQueryParamLimit:
+			uintVal, err := normalizeIntToUint(optionValue)
+			if err != nil {
+				Warn(fmt.Sprintf("%v", err))
+			}
+			viewQuery.Limit(uintVal)
+		case ViewQueryParamDescending:
+			if asBool(optionValue) == true {
+				viewQuery.Order(gocb.Descending)
+			}
+		case ViewQueryParamSkip:
+			uintVal, err := normalizeIntToUint(optionValue)
+			if err != nil {
+				Warn(fmt.Sprintf("%v", err))
+			}
+			viewQuery.Skip(uintVal)
+		case ViewQueryParamGroup:
+			viewQuery.Group(asBool(optionValue))
+		case ViewQueryParamGroupLevel:
+			uintVal, err := normalizeIntToUint(optionValue)
+			if err != nil {
+				Warn(fmt.Sprintf("%v", err))
+			}
+			viewQuery.GroupLevel(uintVal)
+		case ViewQueryParamKey:
+			viewQuery.Key(optionValue)
+		case ViewQueryParamKeys:
+			stringKeys := optionValue.([]string)
+			emptyInterfaceKeys := []interface{}{}
+			for _, key := range stringKeys {
+				emptyInterfaceKeys = append(emptyInterfaceKeys, key)
+			}
+			viewQuery.Keys(emptyInterfaceKeys)
+		}
+
+	}
+
+	// Range: startkey, endkey, inclusiveend
+	var startKey, endKey interface{}
+	if _, ok := params[ViewQueryParamStartKey]; ok {
+		startKey = params[ViewQueryParamStartKey]
+	}
+	if _, ok := params[ViewQueryParamEndKey]; ok {
+		endKey = params[ViewQueryParamEndKey]
+	}
+	inclusiveEnd := false
+	if _, ok := params[ViewQueryParamInclusiveEnd]; ok {
+		inclusiveEnd = asBool(params[ViewQueryParamInclusiveEnd])
+	}
+	viewQuery.Range(startKey, endKey, inclusiveEnd)
+
+	// IdRange: startKeyDocId, endKeyDocId
+	startKeyDocId := ""
+	endKeyDocId := ""
+	if _, ok := params[ViewQueryParamStartKeyDocId]; ok {
+		startKeyDocId = params[ViewQueryParamStartKeyDocId].(string)
+	}
+	if _, ok := params[ViewQueryParamEndKeyDocId]; ok {
+		endKeyDocId = params[ViewQueryParamEndKeyDocId].(string)
+	}
+	viewQuery.IdRange(startKeyDocId, endKeyDocId)
+
+}
+
+func normalizeIntToUint(value interface{}) (uint, error) {
+	switch typeValue := value.(type) {
+	case uint:
+		return typeValue, nil
+	case int:
+		return uint(typeValue), nil
+	case string:
+		i, err := strconv.Atoi(typeValue)
+		return uint(i), err
+	default:
+		return uint(0), fmt.Errorf("Unable to convert %v (%T) -> uint.", value, value)
+	}
+}
+
+func asBool(value interface{}) bool {
+	switch typeValue := value.(type) {
+	case string:
+		switch typeValue {
+		case "true":
+			return true
+		case "false":
+			return false
+		default:
+			Warn("asBool called with unknown value: %v.  defaulting to false", typeValue)
+			return false
+
+		}
+	case bool:
+		return typeValue
+	default:
+		Warn("asBool called with unknown type: %T.  defaulting to false", typeValue)
+		return false
+	}
+
+}
+
+// This prevents Sync Gateway from having too many outstanding concurrent view queries against Couchbase Server
+func (bucket CouchbaseBucketGoCB) waitForAvailViewOp() {
+	bucket.viewOps <- struct{}{}
+	gocbExpvars.Add("ViewOps", 1)
+}
```

Comment:
Yeah I think it follows the "Don't make me think" principle when reading the code.  You see "Wait for available view op" and you know exactly what it does, rather than having to mentally parse a bunch of cryptic stuff like `<- struct{}{}` and see instrumentation noise like `gocbExpvars.Add("ViewOps", 1)`